// @sap/cds/lib/models/2tabledata.js
const DEBUG = /\b(y|all|deploy)\b/.test (process.env.DEBUG) && console.warn // eslint-disable-line
const cds = require ('../cds')
const cdsv = require ('./cdsv')
const CSV = require('../utils/csv')
const util = require ('util')
const readdir = util.promisify (cds.utils.readdir)
const readFile = util.promisify (cds.utils.readFile)
const { path, isdir } = cds.utils

const isCsv = filename => filename[0] !== '-' && filename.endsWith ('.csv')

module.exports.create = async (model, options={}) => {
  const baseDir = options.baseDir  // where the hdbtabledata will be located, for usage in the file_name path
  const hanaModel = cds.reflect(cdsv.toHana (model, {csn: true, src: false}).csn) // need persistence model for the DB names
  const dirs = Array.isArray(options.dirs) ? options.dirs : _csvDirs (model._sources.map (path.dirname))
  if (dirs.length === 0)  return []  // nothing to do

  const datas = (await Promise.all(dirs.map(async dir => {
    let files = []
    if (isdir(dir)) {
      files = await readdir (dir)
    }
    return Promise.all(files.filter (isCsv).map (file => _tabledata4 (dir, file, hanaModel, baseDir)))
  })))
  .reduce((a, b) => a.concat(b), [])
  return _toOutput (datas)
}

async function _tabledata4 (dir, csvFile, model, baseDir)  {
  const baseFileName = path.parse(csvFile).name
  const entityName = baseFileName.replace(/-/g,'.')
  let entity = _entity4 (entityName, model)
  if (!entity) {
    return DEBUG && DEBUG (`[hdbtabledata] - warning: entity ${entityName} not in model`)  // eslint-disable-line
  }

  const tabledata = { format_version: 1, imports: [] }
  const _import = {
    target_table: entity['@cds.persistence.name'],
    source_data: { data_type: 'CSV', file_name: csvFile, has_header: true, type_config: {} },
    import_settings: { import_columns: [], include_filter : [] },
    column_mappings: {}
  };

  let file = path.join(dir, csvFile)
  const src = await readFile (file, 'utf8')
  const { cols, delimiter } = CSV.parseHeader (src)

  cols.forEach(col => {
    const element = _element4 (col, entity)
    if (!element) {  return DEBUG &&
      DEBUG (`[hdbtabledata] - warning: ${col} not in ${entityName}. Elements: ${Object.keys(entity.elements)}`)  // eslint-disable-line
    }
    const tableCol = element['@cds.persistence.name']
    _import.import_settings.import_columns.push (tableCol)
    _import.column_mappings[tableCol] = col
  })

  if (_import.import_settings.import_columns.length === 0)  return  // no columns at all -> skip import
  _import.source_data.type_config.delimiter = delimiter
  tabledata.imports.push (_import)
  return [
    tabledata, {
      file: baseFileName + '.hdbtabledata',
      folder: (baseDir||dir), // as metadata, add the dir to which the csvs are relative to
      csvFolder: dir
    }
  ]
}

function _entity4 (name, csn) {
  const entity = csn.entities [name]
  if (!entity) {
    // text entity -> get base entity
    if (name.endsWith('_texts')) {
        const base = csn.entities [name.slice(0,-6)]
        if (base)  return _entity4 (base.elements.texts.target, csn)
    }
    return null
  }
  // projection entity -> get underlying data entity
  else if (entity.query && entity.query.SELECT && entity.query.SELECT.from && entity.query.SELECT.from.ref) {
    return _entity4 (entity.query.SELECT.from.ref[0], csn)
  }
  return entity
}

function _element4 (name, entity) {
  for (const [elName, element] of Object.entries (entity.elements)) {
    if (elName.toLowerCase() === name.toLowerCase())  return element
  }
}

function _csvDirs (sources) {
  sources = Array.from (new Set(sources)) // uniq
  const folders = []
  for (let src of sources) {
    for (let data of ['/data','/csv']) {
      for (let each of [ src+data, src+'/src'+data, src+'/..'+data ]) {
        let folder = path.resolve (each)
        if (isdir (folder))  folders.push (folder)
      }
    }
  }
  return folders
}

// generator function compliant to what `cds.compile.to` backends can return
function* _toOutput (datas) {
  for (let i = 0; i < datas.length; i++) {
    if (datas[i])  yield datas[i]
  }
}
