// @sap/cds-compiler/lib/gen/csn.js
//csn validator
const H = require("../json/validator/validatorHelper");
function Z(P,CN) { return P.concat("{"+CN+"}");}
let CK=H.checkTypeOf;
let RE=H.reportError;
let FU=H.fuzzy;
let FF={};
function f_0(N,C,P=[]) {
	let CN="csn";
	let PP=["definitions","extensions","version","messages","namespace","meta","$version"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["definitions"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'definitions'")
	}
	if(R) {
		for(let E in N) {
			if (E==="definitions") {
				if(!f_8(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="extensions") {
				if(!f_6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="version") {
				if(!f_3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="messages") {
				if(!f_2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="namespace") {
				if(!f_1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="meta") {
				if(!f_5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$version") {
				if(!f_4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	H.postProcess(N,C,FF,CN);
	return R;
}
FF["csn"]=f_0;
function f_1(N,C,P) {
	let CN="namespace";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c5(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["namespace"]=f_1;
function f_2(N,C,P) {
	let CN="messages";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["messages"]=f_2;
function f_3(N,C,P) {
	let CN="version";
	let PP=["creator","csn"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["csn"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'csn'")
	}
	if(R) {
		for(let E in N) {
			if (E==="creator") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="csn") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["version"]=f_3;
function f_4(N,C,P) {
	let CN="$version";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c5(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["$version"]=f_4;
function f_5(N,C,P) {
	let CN="meta";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["meta"]=f_5;
function f_6(N,C,P) {
	let CN="extensions";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_7(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["extensions"]=f_6;
function f_7(N,C,P) {
	let CN="extension";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["extension"]=f_7;
function f_8(N,C,P) {
	let CN="definitions";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_b(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["definitions"]=f_8;
function mf_definition_0(X) {
	if(X && !X.kind && X.query!==undefined) X.kind="entity";
}
function mf_definition_1(X) {
	if(X && !X.kind && X.type!==undefined) X.kind="type";
}
function mf_definition_2(X) {
	if(X && !X.kind && X.elements!==undefined) X.kind="type";
}
function mf_definition_3(X) {
	if(X && !X.kind) X.kind="entity";
}
function f_b(N,C,P) {
	let CN="definition";
	let PP=["kind","$syntax","localized","params","abstract","actions","includes","query","elements","target","type","keys","cardinality","dbType","items","enum","length","precision","scale","srid","returns"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	mf_definition_0(N);
	mf_definition_1(N);
	mf_definition_2(N);
	mf_definition_3(N);
	let post=f_14(N && N["kind"],C,p);
	if(post) {
		H.hit(C,N,"*elementKind",p.concat(["{elementKind}","kind"]),false);
		post(N,C,p);
	} else {
		R=false;
		H.rmv(C,p.concat(["{elementKind}","kind"]))
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["definition"]=f_b;
function f_d(N,C,P) {
	let CN="items";
	let PP=["elements","type","enum","kind","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_30,rc);
	rc=CK(C,p,N,f_e,rc);
	rc=CK(C,p,N,f_12,rc);
	rc=CK(C,p,N,f_11,rc);
	rc=CK(C,p,N,f_c7,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Elements, enumItems, basicType, typeRef, EmptyObject");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["items"]=f_d;
function f_e(N,C,P) {
	let CN="enumItems";
	let PP=["enum","kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_f,rc);
	rc=CK(C,p,N,f_10,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of enumItemsB, enumItemsR");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["enumItems"]=f_e;
function f_f(N,C,P) {
	let CN="enumItemsB";
	let PP=["enum","kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["enumItemsB"]=f_f;
function f_10(N,C,P) {
	let CN="enumItemsR";
	let PP=["enum","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["enumItemsR"]=f_10;
function f_11(N,C,P) {
	let CN="typeRef";
	let PP=["type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeRef"]=f_11;
function f_12(N,C,P) {
	let CN="basicType";
	let PP=["kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["basicType"]=f_12;
function f_13(N,C,P) {
	let CN="basicTypeAnno";
	let PP=["kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["basicTypeAnno"]=f_13;
function f_14(N,C,P) {
	let CN="elementKind";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="entity")
		R=f_47;
	if(N==="view")
		R=f_48;
	if(N==="annotation")
		R=f_bc;
	if(N==="service")
		R=f_17;
	if(N==="type")
		R=f_1c;
	if(N==="context")
		R=f_15;
	if(N==="action")
		R=f_16;
	if(N==="function")
		R=f_19;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["elementKind"]=f_14;
function f_15(N,C,P) {
	let CN="context";
	let PP=["kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["context"]=f_15;
function f_16(N,C,P) {
	let CN="action";
	let PP=["returns","params","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="returns") {
				if(!f_41(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="params") {
				if(!f_39(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["action"]=f_16;
function f_17(N,C,P) {
	let CN="service";
	let PP=["includes","actions","abstract","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="includes") {
				if(!f_18(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="actions") {
				if(!f_34(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="abstract") {
				if(!f_b8(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["service"]=f_17;
function f_18(N,C,P) {
	let CN="includes";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_c5(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["includes"]=f_18;
function f_19(N,C,P) {
	let CN="function";
	let PP=["returns","params","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["returns"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'returns'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="returns") {
				if(!f_41(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="params") {
				if(!f_39(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["function"]=f_19;
function f_1c(N,C,P) {
	let CN="typeDef";
	let PP=["target","type","keys","cardinality","kind","$syntax","localized","elements","includes","dbType","items","enum","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_1e,rc);
	rc=CK(C,p,N,f_21,rc);
	rc=CK(C,p,N,f_22,rc);
	rc=CK(C,p,N,f_23,rc);
	rc=CK(C,p,N,f_24,rc);
	rc=CK(C,p,N,f_1d,rc);
	rc=CK(C,p,N,f_13,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of typeDefAssoc, typeDefDerived, typeDefStruct, typeDefArray, typeDefEnum, typeDefRef, basicTypeAnno");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["typeDef"]=f_1c;
function f_1d(N,C,P) {
	let CN="typeDefRef";
	let PP=["type","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefRef"]=f_1d;
function f_1e(N,C,P) {
	let CN="typeDefAssoc";
	let PP=["target","type","keys","cardinality","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_1f,rc);
	rc=CK(C,p,N,f_20,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of typeDefDerivedAssoc, typeDefGenericAssoc");
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(N["target"]!==undefined) {
		if(f_c5(N && N["target"],C,p.concat("target")))
			R=true;
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["typeDefAssoc"]=f_1e;
function f_1f(N,C,P) {
	let CN="typeDefDerivedAssoc";
	let PP=["type","target","keys","cardinality","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_2c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="keys") {
				if(!f_27(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefDerivedAssoc"]=f_1f;
function f_20(N,C,P) {
	let CN="typeDefGenericAssoc";
	let PP=["type","target","keys","cardinality","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_2b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="keys") {
				if(!f_27(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefGenericAssoc"]=f_20;
function f_21(N,C,P) {
	let CN="typeDefDerived";
	let PP=["type","keys","cardinality","elements","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_2c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="keys") {
				if(!f_27(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefDerived"]=f_21;
function f_22(N,C,P) {
	let CN="typeDefStruct";
	let PP=["elements","type","kind","$syntax","localized","includes","dbType"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="includes") {
				if(!f_18(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="dbType") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefStruct"]=f_22;
function f_23(N,C,P) {
	let CN="typeDefArray";
	let PP=["items","type","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["items"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'items'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="items") {
				if(!f_d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefArray"]=f_23;
function f_24(N,C,P) {
	let CN="typeDefEnum";
	let PP=["enum","kind","type","length","precision","scale","srid","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_25,rc);
	rc=CK(C,p,N,f_26,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of typeDefEnumBasic, typeDefEnumRef");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["typeDefEnum"]=f_24;
function f_25(N,C,P) {
	let CN="typeDefEnumBasic";
	let PP=["enum","kind","type","length","precision","scale","srid","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefEnumBasic"]=f_25;
function f_26(N,C,P) {
	let CN="typeDefEnumRef";
	let PP=["enum","type","kind","$syntax","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefEnumRef"]=f_26;
function f_27(N,C,P) {
	let CN="arrayOfRefsAs";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_a1(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["arrayOfRefsAs"]=f_27;
function f_28(N,C,P) {
	let CN="cardinality";
	let PP=["min","max","src"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (E==="min") {
				if(!f_bf(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="max") {
				if(!f_29(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="src") {
				if(!f_29(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["cardinality"]=f_28;
function f_29(N,C,P) {
	let CN="UIntegerOrStar";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_bf,rc);
	rc=CK(C,p,N,f_2d,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of UInteger, Star");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["UIntegerOrStar"]=f_29;
function f_2a(N,C,P) {
	let CN="TypeAssocociationOrString";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X==="cds.Association" || X==="cds.Composition" || (typeof X==="string");
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected TypeAssocociationOrString");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["TypeAssocociationOrString"]=f_2a;
function f_2b(N,C,P) {
	let CN="TypeEqAssociation";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X==="cds.Association" || X==="cds.Composition";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected TypeEqAssociation");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["TypeEqAssociation"]=f_2b;
function f_2c(N,C,P) {
	let CN="StringNotAssociation";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => (typeof X==="string") && X!=="cds.Association" && X!=="cds.Composition";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected StringNotAssociation");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["StringNotAssociation"]=f_2c;
function f_2d(N,C,P) {
	let CN="Star";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X==="*";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Star");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Star"]=f_2d;
function f_30(N,C,P) {
	let CN="Elements";
	let PP=["elements","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(R) {
		for(let E in N) {
			if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Elements"]=f_30;
function f_32(N,C,P) {
	let CN="elements";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_96(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["elements"]=f_32;
function f_34(N,C,P) {
	let CN="actions";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_35(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["actions"]=f_34;
function f_35(N,C,P) {
	let CN="functionOrAction";
	let PP=["kind","returns","params"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	let post=f_36(N && N["kind"],C,p);
	if(post) {
		H.hit(C,N,"*functionOrActionKind",p.concat(["{functionOrActionKind}","kind"]),false);
		post(N,C,p);
	} else {
		R=false;
		H.rmv(C,p.concat(["{functionOrActionKind}","kind"]))
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["functionOrAction"]=f_35;
function f_36(N,C,P) {
	let CN="functionOrActionKind";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="function")
		R=f_38;
	if(N==="action")
		R=f_37;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["functionOrActionKind"]=f_36;
function f_37(N,C,P) {
	let CN="anAction";
	let PP=["returns","params","kind"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="returns") {
				if(!f_41(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="params") {
				if(!f_39(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["anAction"]=f_37;
function f_38(N,C,P) {
	let CN="aFunction";
	let PP=["returns","params","kind"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="returns") {
				if(!f_41(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="params") {
				if(!f_39(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["aFunction"]=f_38;
function f_39(N,C,P) {
	let CN="params";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_3a(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["params"]=f_39;
function f_3a(N,C,P) {
	let CN="paramItem";
	let PP=["kind","items","type","elements","ref","length","notNull","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_3f,rc);
	rc=CK(C,p,N,f_3e,rc);
	rc=CK(C,p,N,f_3d,rc);
	rc=CK(C,p,N,f_3b,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of paramArray, paramStruct, paramRef, paramBasic");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["paramItem"]=f_3a;
function f_3b(N,C,P) {
	let CN="paramBasic";
	let PP=["type","kind","length","notNull","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_3c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_40(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="notNull") {
				if(!f_b5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["paramBasic"]=f_3b;
function f_3c(N,C,P) {
	let CN="paramTypeRef";
	let PP=["ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_8a,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of String, Ref");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["paramTypeRef"]=f_3c;
function f_3d(N,C,P) {
	let CN="paramRef";
	let PP=["kind","ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_40(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["paramRef"]=f_3d;
function f_3e(N,C,P) {
	let CN="paramStruct";
	let PP=["kind","elements","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_40(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["paramStruct"]=f_3e;
function f_3f(N,C,P) {
	let CN="paramArray";
	let PP=["kind","items","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(N["items"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'items'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_40(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="items") {
				if(!f_d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["paramArray"]=f_3f;
function f_40(N,C,P) {
	let CN="kindParam";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="param")
		R=f_c5;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["kindParam"]=f_40;
function f_41(N,C,P) {
	let CN="returns";
	let PP=["elements","type","enum","kind","length","precision","scale","srid","items"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_44,rc);
	rc=CK(C,p,N,f_45,rc);
	rc=CK(C,p,N,f_46,rc);
	rc=CK(C,p,N,f_43,rc);
	rc=CK(C,p,N,f_42,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of tElements, tEnum, tItems, tRef, tBasic");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["returns"]=f_41;
function f_42(N,C,P) {
	let CN="tBasic";
	let PP=["kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["tBasic"]=f_42;
function f_43(N,C,P) {
	let CN="tRef";
	let PP=["type","kind","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["tRef"]=f_43;
function f_44(N,C,P) {
	let CN="tElements";
	let PP=["elements","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(R) {
		for(let E in N) {
			if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["tElements"]=f_44;
function f_45(N,C,P) {
	let CN="tEnum";
	let PP=["enum","kind","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["tEnum"]=f_45;
function f_46(N,C,P) {
	let CN="tItems";
	let PP=["items","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["items"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'items'")
	}
	if(R) {
		for(let E in N) {
			if (E==="items") {
				if(!f_d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["tItems"]=f_46;
function f_47(N,C,P) {
	let CN="entity";
	let PP=["params","abstract","actions","includes","kind","$syntax","localized","query","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="params") {
				if(!f_39(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="abstract") {
				if(!f_b8(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="actions") {
				if(!f_34(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="includes") {
				if(!f_18(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="query") {
				if(!f_49(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["entity"]=f_47;
function f_48(N,C,P) {
	let CN="view";
	let PP=["actions","kind","$syntax","localized","query","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["kind"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'kind'")
	}
	if(R) {
		for(let E in N) {
			if (E==="actions") {
				if(!f_34(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_14(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="$syntax") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="query") {
				if(!f_49(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["view"]=f_48;
function f_49(N,C,P) {
	let CN="query";
	let PP=["SELECT","as","SET","op","from","all","kind","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_4c,rc);
	rc=CK(C,p,N,f_4b,rc);
	rc=CK(C,p,N,f_4a,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of querySelect, querySet, queryGen");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["query"]=f_49;
function f_4a(N,C,P) {
	let CN="queryGen";
	let PP=["op","from","all","kind","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(N["op"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'op'")
	}
	if(R) {
		for(let E in N) {
			if (E==="op") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="from") {
				if(!f_8b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="all") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_4d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["queryGen"]=f_4a;
function f_4b(N,C,P) {
	let CN="querySet";
	let PP=["SET","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["SET"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'SET'")
	}
	if(R) {
		for(let E in N) {
			if (E==="SET") {
				if(!f_4e(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["querySet"]=f_4b;
function f_4c(N,C,P) {
	let CN="querySelect";
	let PP=["SELECT","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["SELECT"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'SELECT'")
	}
	if(R) {
		for(let E in N) {
			if (E==="SELECT") {
				if(!f_52(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["querySelect"]=f_4c;
function f_4d(N,C,P) {
	let CN="kindQuery";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="query")
		R=f_49;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["kindQuery"]=f_4d;
function f_4e(N,C,P) {
	let CN="SET";
	let PP=["op","args","orderBy","limit","all"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["args"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'args'")
	}
	if(R) {
		for(let E in N) {
			if (E==="op") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_4f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="orderBy") {
				if(!f_5c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="limit") {
				if(!f_5b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="all") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["SET"]=f_4e;
function f_4f(N,C,P) {
	let CN="qargs";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_50(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["qargs"]=f_4f;
function f_50(N,C,P) {
	let CN="1arg";
	let PP=["SELECT","as","SET","op","from","all","kind","elements","ref","join","args","on"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_49,rc);
	rc=CK(C,p,N,f_a1,rc);
	rc=CK(C,p,N,f_51,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of query, RefAs, JoinOn");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["1arg"]=f_50;
function f_51(N,C,P) {
	let CN="JoinOn";
	let PP=["join","args","on"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["join"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'join'")
	}
	if(N["args"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'args'")
	}
	if(R) {
		for(let E in N) {
			if (E==="join") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_4f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="on") {
				if(!f_65(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["JoinOn"]=f_51;
function f_52(N,C,P) {
	let CN="SELECT";
	let PP=["from","columns","mixin","where","all","orderBy","limit","groupBy","excluding","distinct","having"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["from"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'from'")
	}
	if(R) {
		for(let E in N) {
			if (E==="from") {
				if(!f_8b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="columns") {
				if(!f_6e(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="mixin") {
				if(!f_6c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="where") {
				if(!f_64(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="all") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="orderBy") {
				if(!f_5c(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="limit") {
				if(!f_5b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="groupBy") {
				if(!f_56(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="excluding") {
				if(!f_55(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="distinct") {
				if(!f_b7(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="having") {
				if(!f_53(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["SELECT"]=f_52;
function f_53(N,C,P) {
	let CN="having";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_54(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["having"]=f_53;
function f_54(N,C,P) {
	let CN="havingItem";
	let PP=["ref","val","literal","func","args","as","cast"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_80,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Ref, ValOptLiteral, String, funcCall");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["havingItem"]=f_54;
function f_55(N,C,P) {
	let CN="excluding";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_c5(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["excluding"]=f_55;
function f_56(N,C,P) {
	let CN="groupBy";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_57(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["groupBy"]=f_56;
function f_57(N,C,P) {
	let CN="groupByItem";
	let PP=["ref","xpr","val","literal","func","args","as","cast"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_58,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_80,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Ref, XprGroupBy, ValOptLiteral, funcCall");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["groupByItem"]=f_57;
function f_58(N,C,P) {
	let CN="XprGroupBy";
	let PP=["xpr"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["xpr"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'xpr'")
	}
	if(R) {
		for(let E in N) {
			if (E==="xpr") {
				if(!f_59(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["XprGroupBy"]=f_58;
function f_59(N,C,P) {
	let CN="xprGroupBy";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_5a(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["xprGroupBy"]=f_59;
function f_5a(N,C,P) {
	let CN="xprGroupByItem";
	let PP=["ref","val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_ae,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Ref, String, ValOptLiteral");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["xprGroupByItem"]=f_5a;
function f_5b(N,C,P) {
	let CN="limit";
	let PP=["rows","offset"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(!f_c9(N,C,p,["rows","offset"])) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (E==="rows") {
				if(!f_ab(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="offset") {
				if(!f_ac(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["limit"]=f_5b;
function f_5c(N,C,P) {
	let CN="orderBy";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_5e(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["orderBy"]=f_5c;
function f_5e(N,C,P) {
	let CN="orderByItem";
	let PP=["ref","sort","nulls","xpr","val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_5f,rc);
	rc=CK(C,p,N,f_60,rc);
	rc=CK(C,p,N,f_61,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of sortableRef, sortableXprGroupBy, sortableValOptLiteral");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["orderByItem"]=f_5e;
function f_5f(N,C,P) {
	let CN="sortableRef";
	let PP=["ref","sort","nulls"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="sort") {
				if(!f_63(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="nulls") {
				if(!f_62(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["sortableRef"]=f_5f;
function f_60(N,C,P) {
	let CN="sortableXprGroupBy";
	let PP=["xpr","sort","nulls"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["xpr"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'xpr'")
	}
	if(R) {
		for(let E in N) {
			if (E==="xpr") {
				if(!f_59(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="sort") {
				if(!f_63(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="nulls") {
				if(!f_62(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["sortableXprGroupBy"]=f_60;
function f_61(N,C,P) {
	let CN="sortableValOptLiteral";
	let PP=["val","literal","sort","nulls"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(R) {
		for(let E in N) {
			if (E==="val") {
				if(!f_b0(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="literal") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="sort") {
				if(!f_63(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="nulls") {
				if(!f_62(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["sortableValOptLiteral"]=f_61;
function f_62(N,C,P) {
	let CN="nulls";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="first")
		R=f_c5;
	if(N==="last")
		R=f_c5;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["nulls"]=f_62;
function f_63(N,C,P) {
	let CN="sort";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="asc")
		R=f_c5;
	if(N==="desc")
		R=f_c5;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["sort"]=f_63;
function f_64(N,C,P) {
	let CN="where";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_65(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["where"]=f_64;
function f_65(N,C,P) {
	let CN="condition";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_66(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["condition"]=f_65;
function f_66(N,C,P) {
	let CN="conditionItem";
	let PP=["SELECT","as","SET","op","from","all","kind","elements","val","literal","func","args","cast","param","ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_49,rc);
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_80,rc);
	rc=CK(C,p,N,f_67,rc);
	rc=CK(C,p,N,f_68,rc);
	rc=CK(C,p,N,f_8a,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of query, String, ValOptLiteral, funcCall, paramRefItem, condiRef, Ref");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["conditionItem"]=f_66;
function f_67(N,C,P) {
	let CN="paramRefItem";
	let PP=["param","ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["param"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'param'")
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="param") {
				if(!f_c3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["paramRefItem"]=f_67;
function f_68(N,C,P) {
	let CN="condiRef";
	let PP=["ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_69(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["condiRef"]=f_68;
function f_69(N,C,P) {
	let CN="condiRefArray";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_6a(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["condiRefArray"]=f_69;
function f_6a(N,C,P) {
	let CN="condiRefArrayItem";
	let PP=["id","where","cardinality"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_6b,rc);
	rc=CK(C,p,N,f_c5,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of condiRefO, String");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["condiRefArrayItem"]=f_6a;
function f_6b(N,C,P) {
	let CN="condiRefO";
	let PP=["id","where","cardinality"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["id"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'id'")
	}
	if(N["where"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'where'")
	}
	if(R) {
		for(let E in N) {
			if (E==="id") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="where") {
				if(!f_64(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["condiRefO"]=f_6b;
function f_6c(N,C,P) {
	let CN="mixin";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_6d(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["mixin"]=f_6c;
function f_6d(N,C,P) {
	let CN="mixinItem";
	let PP=["type","target","on","cardinality"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(N["on"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'on'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="on") {
				if(!f_65(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["mixinItem"]=f_6d;
function f_6e(N,C,P) {
	let CN="columns";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_6f(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["columns"]=f_6e;
function f_6f(N,C,P) {
	let CN="column";
	let PP=["ref","as","cast","key","param","func","args","val","literal","SELECT","SET","op","from","all","kind","elements","xpr"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_7a,rc);
	rc=CK(C,p,N,f_70,rc);
	rc=CK(C,p,N,f_75,rc);
	rc=CK(C,p,N,f_af,rc);
	rc=CK(C,p,N,f_71,rc);
	rc=CK(C,p,N,f_86,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of String, complexRef, funcCallColumn, valColumnCast, ValOptLiteralAs, queryCast, xprColumn");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["column"]=f_6f;
function f_70(N,C,P) {
	let CN="funcCallColumn";
	let PP=["key","func","args","as","cast"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["func"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'func'")
	}
	if(R) {
		for(let E in N) {
			if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="func") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_81(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["funcCallColumn"]=f_70;
function f_71(N,C,P) {
	let CN="queryCast";
	let PP=["cast","SELECT","as","SET","op","from","all","kind","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_72,rc);
	rc=CK(C,p,N,f_73,rc);
	rc=CK(C,p,N,f_74,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of querySelectCast, querySetCast, queryGenCast");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["queryCast"]=f_71;
function f_72(N,C,P) {
	let CN="querySelectCast";
	let PP=["cast","SELECT","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["SELECT"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'SELECT'")
	}
	if(R) {
		for(let E in N) {
			if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="SELECT") {
				if(!f_52(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["querySelectCast"]=f_72;
function f_73(N,C,P) {
	let CN="querySetCast";
	let PP=["cast","SET","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["SET"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'SET'")
	}
	if(R) {
		for(let E in N) {
			if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="SET") {
				if(!f_4e(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["querySetCast"]=f_73;
function f_74(N,C,P) {
	let CN="queryGenCast";
	let PP=["cast","op","from","all","kind","elements"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(N["op"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'op'")
	}
	if(R) {
		for(let E in N) {
			if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="op") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="from") {
				if(!f_8b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="all") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="kind") {
				if(!f_4d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["queryGenCast"]=f_74;
function f_75(N,C,P) {
	let CN="valColumnCast";
	let PP=["val","cast","as","key"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(N["cast"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'cast'")
	}
	if(N["as"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'as'")
	}
	if(R) {
		for(let E in N) {
			if (E==="val") {
				if(!f_b0(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["valColumnCast"]=f_75;
function f_76(N,C,P) {
	let CN="castColumn";
	let PP=["target","type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_77,rc);
	rc=CK(C,p,N,f_78,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of castTarget, castType");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["castColumn"]=f_76;
function f_77(N,C,P) {
	let CN="castTarget";
	let PP=["target"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(R) {
		for(let E in N) {
			if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["castTarget"]=f_77;
function f_78(N,C,P) {
	let CN="castType";
	let PP=["type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_79(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["castType"]=f_78;
function f_79(N,C,P) {
	let CN="stringOrRef";
	let PP=["ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_8a,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of String, Ref");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["stringOrRef"]=f_79;
function f_7a(N,C,P) {
	let CN="complexRef";
	let PP=["ref","as","cast","key","param"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_7b(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_76(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="param") {
				if(!f_c3(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["complexRef"]=f_7a;
function f_7b(N,C,P) {
	let CN="complexRefArray";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_7c(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["complexRefArray"]=f_7b;
function f_7c(N,C,P) {
	let CN="complexRefItem";
	let PP=["id","cardinality","where","as","cast","args"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_7d,rc);
	rc=CK(C,p,N,f_c5,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of complexRefItemRef, String");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["complexRefItem"]=f_7c;
function f_7d(N,C,P) {
	let CN="complexRefItemRef";
	let PP=["id","cardinality","where","as","cast","args"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["id"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'id'")
	}
	if(R) {
		for(let E in N) {
			if (E==="id") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="where") {
				if(!f_64(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_7e(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["complexRefItemRef"]=f_7d;
function f_7e(N,C,P) {
	let CN="riArgs";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_ae(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["riArgs"]=f_7e;
function f_7f(N,C,P) {
	let CN="cast";
	let PP=["type","length","precision","scale","srid"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["cast"]=f_7f;
function f_80(N,C,P) {
	let CN="funcCall";
	let PP=["func","args","as","cast"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["func"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'func'")
	}
	if(R) {
		for(let E in N) {
			if (E==="func") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_81(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["funcCall"]=f_80;
function f_81(N,C,P) {
	let CN="funcArgs";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_82,rc);
	rc=CK(C,p,N,f_83,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of funcArgsNamed, funcArgsArray");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["funcArgs"]=f_81;
function f_82(N,C,P) {
	let CN="funcArgsNamed";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_ae(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["funcArgsNamed"]=f_82;
function f_83(N,C,P) {
	let CN="funcArgsArray";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_84(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["funcArgsArray"]=f_83;
function f_84(N,C,P) {
	let CN="funcArg";
	let PP=["ref","xpr","func","args","as","cast","val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_85,rc);
	rc=CK(C,p,N,f_80,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_2d,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Ref, xprArg, funcCall, ValOptLiteral, Star");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["funcArg"]=f_84;
function f_85(N,C,P) {
	let CN="xprArg";
	let PP=["xpr"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["xpr"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'xpr'")
	}
	if(R) {
		for(let E in N) {
			if (E==="xpr") {
				if(!f_87(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["xprArg"]=f_85;
function f_86(N,C,P) {
	let CN="xprColumn";
	let PP=["xpr","as","cast","key"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["xpr"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'xpr'")
	}
	if(R) {
		for(let E in N) {
			if (E==="xpr") {
				if(!f_87(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cast") {
				if(!f_7f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["xprColumn"]=f_86;
function f_87(N,C,P) {
	let CN="xprArrayQuery";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_88(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["xprArrayQuery"]=f_87;
function f_88(N,C,P) {
	let CN="xprArrayQueryItem";
	let PP=["SELECT","as","SET","op","from","all","kind","elements","ref","param","val","literal","func","args","cast"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_49,rc);
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_89,rc);
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_80,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of query, String, refParam, Ref, ValOptLiteral, funcCall");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["xprArrayQueryItem"]=f_88;
function f_89(N,C,P) {
	let CN="refParam";
	let PP=["ref","param"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(N["param"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'param'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="param") {
				if(!f_c3(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["refParam"]=f_89;
function f_8a(N,C,P) {
	let CN="Ref";
	let PP=["ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Ref"]=f_8a;
function f_8b(N,C,P) {
	let CN="from";
	let PP=["ref","as","join","args","on","SELECT","SET"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_8c,rc);
	rc=CK(C,p,N,f_8d,rc);
	rc=CK(C,p,N,f_4c,rc);
	rc=CK(C,p,N,f_4b,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of From, FromJoin, querySelect, querySet");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["from"]=f_8b;
function f_8c(N,C,P) {
	let CN="From";
	let PP=["ref","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["From"]=f_8c;
function f_8d(N,C,P) {
	let CN="FromJoin";
	let PP=["join","args","on"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["join"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'join'")
	}
	if(N["args"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'args'")
	}
	if(R) {
		for(let E in N) {
			if (E==="join") {
				if(!f_8e(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="args") {
				if(!f_4f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="on") {
				if(!f_65(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["FromJoin"]=f_8d;
function f_8e(N,C,P) {
	let CN="join";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="inner")
		R=f_c5;
	if(N==="full")
		R=f_c5;
	if(N==="cross")
		R=f_c5;
	if(N==="left")
		R=f_c5;
	if(N==="right")
		R=f_c5;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["join"]=f_8e;
function f_8f(N,C,P) {
	let CN="ref";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_90(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["ref"]=f_8f;
function f_90(N,C,P) {
	let CN="1ref";
	let PP=["args","id","where"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_92,rc);
	rc=CK(C,p,N,f_95,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of String, idWithArgs, idWithWhere");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["1ref"]=f_90;
function f_92(N,C,P) {
	let CN="idWithArgs";
	let PP=["args","id"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["args"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'args'")
	}
	if(N["id"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'id'")
	}
	if(R) {
		for(let E in N) {
			if (E==="args") {
				if(!f_93(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="id") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["idWithArgs"]=f_92;
function f_93(N,C,P) {
	let CN="idArgs";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_94(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["idArgs"]=f_93;
function f_94(N,C,P) {
	let CN="idArgsItem";
	let PP=["val","literal","param","ref"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_67,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of ValOptLiteral, paramRefItem");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["idArgsItem"]=f_94;
function f_95(N,C,P) {
	let CN="idWithWhere";
	let PP=["where","id"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["where"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'where'")
	}
	if(N["id"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'id'")
	}
	if(R) {
		for(let E in N) {
			if (E==="where") {
				if(!f_64(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="id") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["idWithWhere"]=f_95;
function f_96(N,C,P) {
	let CN="element";
	let PP=["target","on","type","key","keys","cardinality","notNull","items","elements","length","default","enum","precision","scale","srid","virtual","masked","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_9e,rc);
	rc=CK(C,p,N,f_9c,rc);
	rc=CK(C,p,N,f_9a,rc);
	rc=CK(C,p,N,f_99,rc);
	rc=CK(C,p,N,f_9b,rc);
	rc=CK(C,p,N,f_9d,rc);
	rc=CK(C,p,N,f_98,rc);
	rc=CK(C,p,N,f_97,rc);
	rc=CK(C,p,N,f_c7,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of associationElement, arrayElement, structElementRef, structElement, enumElement, typeRefElement, typeDefElement, annotatedOnlyElement, EmptyObject");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["element"]=f_96;
function f_97(N,C,P) {
	let CN="annotatedOnlyElement";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["annotatedOnlyElement"]=f_97;
function f_98(N,C,P) {
	let CN="typeDefElement";
	let PP=["type","key","length","default","precision","scale","srid","notNull","virtual","masked","localized"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="default") {
				if(!f_a7(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="precision") {
				if(!f_b2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="scale") {
				if(!f_b3(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="srid") {
				if(!f_b4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="notNull") {
				if(!f_b5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="virtual") {
				if(!f_ba(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="masked") {
				if(!f_bb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="localized") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeDefElement"]=f_98;
function f_99(N,C,P) {
	let CN="structElement";
	let PP=["elements","key","notNull"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(R) {
		for(let E in N) {
			if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="notNull") {
				if(!f_b5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["structElement"]=f_99;
function f_9a(N,C,P) {
	let CN="structElementRef";
	let PP=["elements","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["elements"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'elements'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="elements") {
				if(!f_32(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["structElementRef"]=f_9a;
function f_9b(N,C,P) {
	let CN="enumElement";
	let PP=["type","length","default","enum"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["enum"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'enum'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="default") {
				if(!f_a8(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="enum") {
				if(!f_a2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["enumElement"]=f_9b;
function f_9c(N,C,P) {
	let CN="arrayElement";
	let PP=["type","key","items"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["items"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'items'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_b6(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="items") {
				if(!f_d(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["arrayElement"]=f_9c;
function f_9d(N,C,P) {
	let CN="typeRefElement";
	let PP=["length","type"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(R) {
		for(let E in N) {
			if (E==="length") {
				if(!f_b1(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_8a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["typeRefElement"]=f_9d;
function f_9e(N,C,P) {
	let CN="associationElement";
	let PP=["target","on","type","key","keys","cardinality","notNull"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_a0,rc);
	rc=CK(C,p,N,f_9f,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of unmanagedAssociationElement, managedAssociationElement");
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(R) {
		for(let E in N) {
			if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(N["target"]!==undefined) {
		if(f_c5(N && N["target"],C,p.concat("target")))
			R=true;
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["associationElement"]=f_9e;
function f_9f(N,C,P) {
	let CN="managedAssociationElement";
	let PP=["type","target","key","keys","cardinality","notNull"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(R) {
		for(let E in N) {
			if (E==="type") {
				if(!f_2a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="keys") {
				if(!f_27(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="notNull") {
				if(!f_b5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["managedAssociationElement"]=f_9f;
function f_a0(N,C,P) {
	let CN="unmanagedAssociationElement";
	let PP=["on","type","target","key","keys","cardinality","notNull"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["on"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'on'")
	}
	if(N["type"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'type'")
	}
	if(N["target"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'target'")
	}
	if(R) {
		for(let E in N) {
			if (E==="on") {
				if(!f_a5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="type") {
				if(!f_2a(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="target") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="key") {
				if(!f_c2(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="keys") {
				if(!f_27(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="cardinality") {
				if(!f_28(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="notNull") {
				if(!f_b5(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(N["on"]!==undefined) {
		if(f_a5(N && N["on"],C,p.concat("on")))
			R=true;
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["unmanagedAssociationElement"]=f_a0;
function f_a1(N,C,P) {
	let CN="RefAs";
	let PP=["ref","as"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["ref"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'ref'")
	}
	if(R) {
		for(let E in N) {
			if (E==="ref") {
				if(!f_8f(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["RefAs"]=f_a1;
function f_a2(N,C,P) {
	let CN="enum";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bd(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_a3(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["enum"]=f_a2;
function f_a3(N,C,P) {
	let CN="1enum";
	let PP=["kind","val"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (E==="kind") {
				if(!f_a4(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="val") {
				if(!f_b0(N && N[E],C,p.concat(E)))
					R=false;
			} else if (f_ce(E,C,p.concat(E),true)) {
				if(!f_cb(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["1enum"]=f_a3;
function f_a4(N,C,P) {
	let CN="kindEnum";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=null;
	if(!f_c5(N,C,p)) {
		R=null;
		H.rmv(C,p);
	}
	if(N==="enum")
		R=f_a2;
	if(R===null)
		RE(C,p,N,"Unexpected property value: "+N+" for "+CN);
	if(R!==null)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["kindEnum"]=f_a4;
function f_a5(N,C,P) {
	let CN="associationCondition";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c8(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R) {
		for(let E in N) {
			if (!f_a6(N && N[E],C,p.concat(E))) {
				R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["associationCondition"]=f_a5;
function f_a6(N,C,P) {
	let CN="associationConditionItem";
	let PP=["ref","val","literal","func","args","as","cast","#"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_68,rc);
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_ae,rc);
	rc=CK(C,p,N,f_c5,rc);
	rc=CK(C,p,N,f_80,rc);
	rc=CK(C,p,N,f_a8,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of condiRef, Ref, ValOptLiteral, String, funcCall, enumVal");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["associationConditionItem"]=f_a6;
function f_a7(N,C,P) {
	let CN="default";
	let PP=["#","ref","val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_a8,rc);
	rc=CK(C,p,N,f_8a,rc);
	rc=CK(C,p,N,f_ae,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of enumVal, Ref, ValOptLiteral");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["default"]=f_a7;
function f_a8(N,C,P) {
	let CN="enumVal";
	let PP=["#"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["#"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property '#'")
	}
	if(R) {
		for(let E in N) {
			if (E==="#") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["enumVal"]=f_a8;
function f_aa(N,C,P) {
	let CN="UIntegerNull";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_bf,rc);
	rc=CK(C,p,N,f_cc,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of UInteger, Null");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["UIntegerNull"]=f_aa;
function f_ab(N,C,P) {
	let CN="UIntNullVal";
	let PP=["val"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(R) {
		for(let E in N) {
			if (E==="val") {
				if(!f_aa(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["UIntNullVal"]=f_ab;
function f_ac(N,C,P) {
	let CN="UIntVal";
	let PP=["val"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(R) {
		for(let E in N) {
			if (E==="val") {
				if(!f_bf(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["UIntVal"]=f_ac;
function f_ae(N,C,P) {
	let CN="ValOptLiteral";
	let PP=["val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(R) {
		for(let E in N) {
			if (E==="val") {
				if(!f_b0(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="literal") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["ValOptLiteral"]=f_ae;
function f_af(N,C,P) {
	let CN="ValOptLiteralAs";
	let PP=["as","val","literal"];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(N["val"]===undefined) {
		R=false;
		RE(C,p,N,"Missing property 'val'")
	}
	if(R) {
		for(let E in N) {
			if (E==="as") {
				if(!f_b9(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="val") {
				if(!f_b0(N && N[E],C,p.concat(E)))
					R=false;
			} else if (E==="literal") {
				if(!f_c5(N && N[E],C,p.concat(E)))
					R=false;
			}
		}
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["ValOptLiteralAs"]=f_af;
function f_b0(N,C,P) {
	let CN="AnyBasic";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let rc=false;
	rc=CK(C,p,N,f_be,rc);
	rc=CK(C,p,N,f_c1,rc);
	rc=CK(C,p,N,f_c4,rc);
	rc=CK(C,p,N,f_c2,rc);
	rc=CK(C,p,N,f_cc,rc);
	if(!rc) {
		R=false;
		RE(C,p,N,CN+": Expected one of Integer, Number, AnyString, Boolean, Null");
	}
	if(R===true)
		H.hit(C,N,CN,p);
	return R;
}
FF["AnyBasic"]=f_b0;
function f_b1(N,C,P) {
	let CN="length";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bf(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["length"]=f_b1;
function f_b2(N,C,P) {
	let CN="precision";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bf(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["precision"]=f_b2;
function f_b3(N,C,P) {
	let CN="scale";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bf(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["scale"]=f_b3;
function f_b4(N,C,P) {
	let CN="srid";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_bf(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["srid"]=f_b4;
function f_b5(N,C,P) {
	let CN="notNull";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["notNull"]=f_b5;
function f_b6(N,C,P) {
	let CN="key";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["key"]=f_b6;
function f_b7(N,C,P) {
	let CN="distinct";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["distinct"]=f_b7;
function f_b8(N,C,P) {
	let CN="abstract";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["abstract"]=f_b8;
function f_b9(N,C,P) {
	let CN="as";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c5(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["as"]=f_b9;
function f_ba(N,C,P) {
	let CN="virtual";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["virtual"]=f_ba;
function f_bb(N,C,P) {
	let CN="masked";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c2(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["masked"]=f_bb;
function f_bc(N,C,P) {
	let CN="annotation";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["annotation"]=f_bc;
function f_bd(N,C,P) {
	let CN="Dict";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	if(!f_c6(N,C,p)) {
		R=false;
		H.rmv(C,p);
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Dict"]=f_bd;
function f_be(N,C,P) {
	let CN="Integer";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "number" && (X == "0" || (X % 1 == 0));
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Integer");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Integer"]=f_be;
function f_bf(N,C,P) {
	let CN="UInteger";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "number" && (X == "0" || (X > 0 && X % 1 == 0));
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected UInteger");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["UInteger"]=f_bf;
function f_c1(N,C,P) {
	let CN="Number";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "number";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Number");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Number"]=f_c1;
function f_c2(N,C,P) {
	let CN="Boolean";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "boolean";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Boolean");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Boolean"]=f_c2;
function f_c3(N,C,P) {
	let CN="True";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X === true;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected True");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["True"]=f_c3;
function f_c4(N,C,P) {
	let CN="AnyString";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "string";
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected AnyString");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["AnyString"]=f_c4;
function f_c5(N,C,P) {
	let CN="String";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === "string" && X.length > 0;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected String");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["String"]=f_c5;
function f_c6(N,C,P) {
	let CN="Object";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === 'object' && X !==null && !Array.isArray(X);
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Object");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Object"]=f_c6;
function f_c7(N,C,P) {
	let CN="EmptyObject";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === 'object' && X !==null && !Array.isArray(X) && Object.keys(X).length===0;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected EmptyObject");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["EmptyObject"]=f_c7;
function f_c8(N,C,P) {
	let CN="Array";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => typeof X === 'object' && X!==null && Array.isArray(X);
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Array");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Array"]=f_c8;
function f_c9(N,C,P,args) {
	let CN="AnyOf";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => Object.keys(X).length>=1;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected AnyOf" + ": " + args.join(", "));
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["AnyOf"]=f_c9;
function f_cb(N,C,P) {
	let CN="Any";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = () => true;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Any");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Any"]=f_cb;
function f_cc(N,C,P) {
	let CN="Null";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X === null;
	if(F(N)===false) {
		R=false;
		RE(C,p,N,"Expected Null");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Null"]=f_cc;
function f_ce(N,C,P,optional) {
	let CN="Annotation";
	let PP=[];
	let p=Z(P,CN);
	if(C.fuzzy)
		return FU(C,p,N,PP,CN);
	let R=true;
	let F = (X) => X.startsWith("@");
	if(F(N)===false) {
		R=false;
		if(!optional)
			RE(C,p,N,"Expected Annotation");
	}
	if(R===true)
		H.hit(C,N,CN,p,true);
	return R;
}
FF["Annotation"]=f_ce;
module.exports = f_0
