// @sap/cds/lib/env/defaults.js
module.exports = {

  features: {
    localized: true
  },

  folders: {
    app: 'app',
    srv: 'srv',
    db: 'db',
  },

  i18n: {
    folders: ['_i18n', 'i18n', 'assets/i18n'],
    for_sqlite: ['de','fr'],
    default_language: 'en'
  },

  requires: {
    db: undefined,

    auth: undefined,
    monitoring: undefined,
    logging: undefined,
    destinations: undefined,
    audit: undefined,
    sql: {
      "[development]": { kind:'sqlite', credentials: { database:':memory:' } },
      "[production]": { kind:'hana' },
    },
    messaging: {
      "[development]": { kind:"local" },
      "[production]": {},
    },
    "enterprise-messaging": {}
  },

  odata: {
    version:"v2"
  },

  build: {
    target: 'gen'
  },

  deploy: {

  },

  mtx: {
    api: {
      model: true,
      provisioning: true,
      metadata: true
    },
    domain: '__default__'
  },

  cdsc: {
    // toSql: { associations: "joins" },
    // newCsn: true,
    smart: {
      // to_many: true
    }
  },

  sql_mapping:"plain", // or 'hdbcds',
}
